Fix/update various type checks

adeak opened the pull request:
### Overview

I've noticed several instances of type checks against `collections.Iterable`. A lot of these should actually be stricter checks, and they should all be using `collections.abc` going forward.

### Details

The first commit replaces every instance of `collections.Iterable` with `collections.abc.Iterable`, as the former is deprecated.

The second commit replaces some instances of `collections.abc.Iterable` checks with `(np.ndarray, collections.abc.Sequence)` checks. The two cases that made this necessary were

  * subsequent `len` checks on the iterables,
  * subsequent subscription of the iterables, and
  * subsequent calls of `np.asarray` on the iterables.

For the above [an `Iterable` is insufficient](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes). For instance `obj = iter([2, 3, 4])` would pass the `Iterable` check, but it would fail both at `len(obj)` and `obj[0]`, and produce an unexpected (object-dtype) array when passed to `np.asarray`. For `len` it would be enough to use `collections.abc.Collection`, but these don't necessarily support subscription (e.g. `set`s). In every case it seemed to me that the collections in question actually had inherent order to them: bounds with from/to, ranges as min/max, things like that. So semantically speaking I believe `collections.abc.Sequence` is the correct abstract base type to check.

The small blemish is that [`numpy.ndarray` doesn't register as a `Sequence`](https://github.com/numpy/numpy/issues/2776), so we also have to explicitly check for `ndarray`s as well (which are a very common use case for most inputs).

### TODO

- [x] also handle `numpy_to_idarr`
- [x] also handle similar instances of `np.asarray`, if any 
- [x] add tests for each fixed `Iterable` case
- [ ] ~maybe do the same for high-profile entry points with array-like inputs~

akaszynski commented on pyvista/core/filters.py:
```diff
@@ -130,13 +130,13 @@ def clip_box(dataset, bounds=None, invert=True, factor=0.35):
         If no bounds are given, a corner of the dataset bounds will be removed.
 
-        Parameters
+        Parametersequence
```
This seems to be a mistake.

akaszynski commented on pyvista/core/filters.py:
```diff
@@ -130,13 +130,13 @@ def clip_box(dataset, bounds=None, invert=True, factor=0.35):
         If no bounds are given, a corner of the dataset bounds will be removed.
 
-        Parameters
+        Parametersequence
```
```suggestion
        Parameters
```

adeak replied:
Ugh, thanks for catching that, I have no idea how that happened :disappointed:

akaszynski commented on pyvista/utilities/helpers.py:
```diff
@@ -505,7 +505,7 @@ def is_inside_bounds(point, bounds):
     if isinstance(point, (int, float)):
         point = [point]
-    if isinstance(point, collections.Iterable) and not isinstance(point, collections.deque):
+    if isinstance(point, collections.abc.Iterable) and not isinstance(point, collections.deque):
```
Any reason why we're not using a `Sequence` here?

adeak replied:
The only reason is that I didn't notice the `len(point)` check... sorry for being sloppy. I thought that it was only passed on to `deque`, which would be happy with an iterable. Although semantically speaking even without the `len` it should be preferable to have a sequence.

akaszynski commented:
Thanks for fixing the depreciation.  As for `Sequence` vs `Iterable`, that's an edge case I had not considered and it's a great catch!

In this PR, would you mind also introducing some small unit tests with invalid `iter` inputs so we don't run into this issue in the future with any potential code base changes?

Finally, there are a ton of places within `pyvista` where we do not perform type checking or conversion and it's quite inconsistent.  If you want, you can add that here or consider helping out in another PR.  We don't need to do it everywhere, but in places we're most likely to encounter it as an issue (e.g. like creating PolyData point clouds from a nested list of points).

adeak commented:
Sure thing, I'll try to come up with some regression tests. I think I won't include other typecheck additions in this PR, mostly because I'm not very familiar with the codebase yet and it would be hard to figure out which parts matter most. With the changes in this PR I have a handful of places to investigate, in order to see how crucial they might be.

akaszynski commented:
I just fixed the conflict.  Would you mind taking a look at `numpy_to_idarr` and implementing the type checking there?

adeak commented:
I didn't know you could push to my branch, nifty, thanks!

Yeah, I'll tackle that too, thanks for the pointer. I haven't decided yet what the best approach will be; probably converting ahead of time with `asarray` and inspecting the `dtype` for faulty cases (probably filtering out `dtype=object`, but we might want to do even more). I think I'll be able to work on it tonight, in a few hours.

akaszynski commented:
Thanks.  Also, please add [wip] or the work in progress icon to the name of the PR so I know not to review and merge it, and then remove it when you're ready to merge.

adeak commented:
Right, thanks. I've already discovered how that works (partly because you suggested that I open a WIP PR for a feature, and partly because I looked at CI and saw the WIP job). Done.

adeak commented:
@akaszynski I've chosen an approach for `numpy_to_idarr` and I'd like to know what you think.

Since `np.asarray` will accept literally anything, I've decided to replace the try/except with a dtype check. For the time being I've allowed `(np.bool_, np.integer, np.floating)` as valid types. Two potential changes to this choice:

  1. don't allow `np.floating`, since we're going to convert to something like `np.intp` anyway, and in numpy native only integer arrays are allowed as indices (as annoying as this might be at times)
  2. also allow `np.complexfloating` as an extreme measure for allowing numerical inputs. Seemed like a stretch to me, that's why I only added floats (with a question mark)

As an aside I've changed `np.bool` to `np.bool_` on a nearby line. `np.bool`, `np.int` and `np.float` are actually aliases to the eponymous native python types. I can see 24 instances of `np.bool`, and while this should work in every appearing context, they should either be `bool` or `np.bool_`. I can fix these too, but this should probably be a different PR :)

akaszynski commented:
> For the time being I've allowed `(np.bool_, np.integer, np.floating)` as valid types

As `numpy_to_idarr` is specifically for indices, we should probably just allow `np.bool_` and `np.integer`.  Afterall, you can't run `a = [1, 2]; a[1.0]`.

> If we agree on a solution for this case I can look at the other instances of `np.asarray` to see if they need the same.

I'm happy with that, though we should be more careful about the allowable types.  I think `vtkIdTypeArray` is a special case, but there might be other special vtk types..  I'll be sure to check through when you're ready for me to review it.


> As an aside I've changed `np.bool` to `np.bool_` on a nearby line. `np.bool`, `np.int` and `np.float` are actually aliases to the eponymous native python types. I can see 24 instances of `np.bool`, and while this should work in every appearing context, they should either be `bool` or `np.bool_`. I can fix these too, but this should probably be a different PR :)

Totally didn't even realize this was an issue, and I think using the `np.bool_` dtype is a bit clearer since we're specifically referencing the "bool as a byte" rather than the python type ([whose size depends on the platform](https://stackoverflow.com/a/32763873/3369879)).

I think a separate PR would be best for this.

adeak commented:
@akaszynski I've done what I could for now :) I haven't looked into other, unguarded entry points now, but I'm already worried that this PR has grown a bit too large. I've fixed a few things that I came across near the iterable checks, but I tried staying near the original goal (I also have a long list of things I think could be tackled). I'll un-WIP this to get it moving, and we can get back to the rest later (but master is also a bit of a moving target :)

I've changed the `Iterable` checks to `Sequence` where necessary. I've also added (mostly failing) test cases to the things I've changed. Some remarks in addition to this:
  1. I've also added checks and a test function for `BasePlotter`'s shape kwarg
  2. During this I decided to convert `shape` to `ndarray` (makes checks easier and shorter), but assign it back to an attribute as a tuple (which is also something that's assumed elsewhere, e.g. https://github.com/pyvista/pyvista/blob/master/pyvista/plotting/plotting.py#L2461)
  3. I've added type checks and a test function for `BasePlotter`'s `loc_to_index` and `index_to_loc`. Might also affect https://github.com/pyvista/pyvista/issues/626 which I should get around too.
  4. Changing `Iterable` to `Sequence` is actually insufficient: strings still pass that (and I'm pretty sure [there are test cases](https://github.com/pyvista/pyvista/blob/master/tests/test_plotting.py#L683-L684) that are only caught because a different `TypeError` is triggered elsewhere). But at least passing a string-valued sequence instead of a numerical list is much less realistic than passing a set or a generator containing numbers. So I left this alone for now.

adeak commented:
Thank you @akaszynski!
